src/lex/replace.rs

```diff
@@ -166,56 +166,104 @@ pub fn replace(
+                data: Token::Id(id),
+                ..
+            }) => {
+                if !ids_seen.contains(&id) {
+                    match definitions.get(&id) {
+                        Some(Definition::Object(replacement_list)) => {
+                            ids_seen.insert(id);
+                            // prepend the new tokens to the pending tokens
+                            // They need to go before, not after. For instance:
+                            // ```c
+                            // #define a b c d
+                            // #define b 1 + 2
+                            // a
+                            // ```
+                            // should replace to `1 + 2 c d`, not `c d 1 + 2`
+                            let mut new_pending = VecDeque::new();
+                            // we need a `clone()` because `self.definitions` needs to keep its copy of the definition
+                            new_pending.extend(
+                                replacement_list
+                                    .iter()
+                                    .cloned()
+                                    .map(|t| Ok(location.with(t))),
+                            );
+                            new_pending.append(&mut pending);
+                            pending = new_pending;
+                            continue;
+                        }
+                        // TODO: so many allocations :(
+                        Some(Definition::Function { .. }) => {
+                            ids_seen.insert(id);
+                            let func_replacements = replace_function(
+                                definitions,
+                                id,
+                                location,
+                                &mut pending,
+                                &mut inner,
+                            );
+                            let mut func_replacements: VecDeque<_> =
+                                func_replacements.into_iter().collect();
+                            func_replacements.append(&mut pending);
+                            pending = func_replacements;
+                            continue;
+                        }
+                        None => {}
                     }
-                    None => {}
                 }
             }
+            Ok(Locatable {
+                data: Token::HashHash(true),
+                ..
+            }) => {
+                let preceding_tok = loop {
+                    match replacements.pop() {
+                        Some(Ok(Locatable {
+                            data: Token::Whitespace(_),
+                            ..
+                        })) => continue,
+                        Some(Ok(Locatable { data: token, .. })) => break token,
+                        None | Some(Err(_)) => {
+                            return wrap_error(&location, CppError::HashHashMissingParameter(true))
+                        }
```

Comment:
Hmm, in the `Some(Err)` case I think we should return the original error instead of swallowing it.